File: app/src/main/java/space/pal/sig/view/apod/DownloadImageWorker.java
package space.pal.sig.view.apod;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Environment;

import androidx.annotation.NonNull;
import androidx.work.Data;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.security.SecureRandom;

import space.pal.sig.service.notification.SpaceNotificationManager;

import static space.pal.sig.view.apod.ImageViewModel.IMAGE_URL;

/**
 * SpaceNow
 * Created by Catalin on 7/28/2020
 **/
public class DownloadImageWorker extends Worker {

    private final Context context;

    public DownloadImageWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
        this.context = context;
    }

    @NonNull
    @Override
    public Result doWork() {
        int notificationId = new SecureRandom().nextInt();
        SpaceNotificationManager.createNotification(context,
                "Downloading image...", notificationId);
        Data data = getInputData();
        String url = data.getString(IMAGE_URL);
        try {
            Bitmap bitmap = downloadFile(url);
            saveImageToGallery(bitmap);
            SpaceNotificationManager.createBigNotification(context,
                    "Image downloaded", notificationId, bitmap);
        } catch (IOException e) {
            e.printStackTrace();
            SpaceNotificationManager.createNotification(context,
                    "Download failed", notificationId);
            return Result.failure();
        }
        return Result.success();
    }

    private Bitmap downloadFile(String urlString) throws IOException {
        Bitmap bitmap = null;
        URL url = new URL(urlString);
        HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
        int statusCode = httpURLConnection.getResponseCode();
        if (statusCode == 200) {
            InputStream inputStream = httpURLConnection.getInputStream();
            if (inputStream != null) {
                bitmap = BitmapFactory.decodeStream(inputStream);
            }
        }
        httpURLConnection.disconnect();
        return bitmap;
    }

    private void saveImageToGallery(Bitmap bitmap) throws IOException {
        String name = String.valueOf(System.currentTimeMillis());
        String fileName = String.format("%s.jpg", name);
        File image = new File(context.getExternalFilesDir(
                Environment.DIRECTORY_PICTURES), fileName);
        OutputStream outStream = new FileOutputStream(image);
        bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outStream);
        outStream.flush();
        outStream.close();
        refreshGallery(image);
    }

    private void refreshGallery(File file) {
        Intent intent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
        intent.setData(Uri.fromFile(file));
        context.sendBroadcast(intent);
    }
}

File: app/src/main/java/space/pal/sig/service/download/DownloadImageWorker.kt
package space.pal.sig.service.download

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Environment
import androidx.work.Worker
import androidx.work.WorkerParameters
import space.pal.sig.service.notification.SpaceNotificationManager.createBigNotification
import space.pal.sig.service.notification.SpaceNotificationManager.createNotification
import space.pal.sig.view.apod.ApodViewModel.Companion.IMAGE_URL
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.security.SecureRandom

/**
SpaceNow
Created by Catalin on 12/27/2020
 **/
class DownloadImageWorker(
        context: Context,
        workerParameters: WorkerParameters
) : Worker(context, workerParameters) {

    override fun doWork(): Result {
        val notificationId = SecureRandom().nextInt()
        createNotification(applicationContext,
                "Downloading image...", notificationId)
        val data = inputData
        val url = data.getString(IMAGE_URL)
        try {
            url?.let {
                val bitmap: Bitmap? = downloadFile(url)
                bitmap?.let {
                    saveImageToGallery(bitmap)
                    createBigNotification(applicationContext,
                            "Image downloaded", notificationId, bitmap)
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
            createNotification(applicationContext,
                    "Download failed", notificationId)
            return Result.failure()
        }
        return Result.success()
    }

    @Throws(IOException::class)
    private fun downloadFile(urlString: String): Bitmap? {
        var bitmap: Bitmap? = null
        val url = URL(urlString)
        val httpURLConnection = url.openConnection() as HttpURLConnection
        val statusCode = httpURLConnection.responseCode
        if (statusCode == 200) {
            val inputStream = httpURLConnection.inputStream
            if (inputStream != null) {
                bitmap = BitmapFactory.decodeStream(inputStream)
            }
        }
        httpURLConnection.disconnect()
        return bitmap
    }

    @Throws(IOException::class)
    private fun saveImageToGallery(bitmap: Bitmap) {
        val name = System.currentTimeMillis().toString()
        val fileName = String.format("%s.jpg", name)
        val image = File(applicationContext.getExternalFilesDir(
                Environment.DIRECTORY_PICTURES), fileName)
        val outStream: OutputStream = FileOutputStream(image)
        bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outStream)
        outStream.flush()
        outStream.close()
        refreshGallery(image)
    }

    private fun refreshGallery(file: File) {
        val intent = Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE)
        intent.data = Uri.fromFile(file)
        applicationContext.sendBroadcast(intent)
    }
}